packages/cojson/src/coValueState.ts

```diff
@@ -89,29 +91,166 @@ type CoValueStateAction =
     };
 
 export class CoValueState {
-  constructor(public state: CoValueUnknownState | CoValueAvailableState) {}
+  value: Promise<CoValueCore | "unavailable">;
+  private resolve: (value: CoValueCore | "unavailable") => void = () => {};
+
+  constructor(
+    public id: RawCoID,
+    public state:
+      | CoValueUnknownState
+      | CoValueLoadingState
+      | CoValueAvailableState
+      | CoValueUnavailableState,
+  ) {
+    if (state.type !== "available") {
+      const { promise, resolve } = createResolvablePromise<
+        CoValueCore | "unavailable"
+      >();
 
-  static Unknown(peersToWaitFor: Set<PeerID>) {
-    return new CoValueState(new CoValueUnknownState(peersToWaitFor));
+      this.value = promise;
+      this.resolve = resolve;
+    } else {
+      this.value = Promise.resolve(state.coValue);
+    }
+  }
+
+  static Unknown(id: RawCoID) {
+    return new CoValueState(id, new CoValueUnknownState());
+  }
+
+  static Loading(id: RawCoID, peersIds: Iterable<PeerID>) {
+    return new CoValueState(id, new CoValueLoadingState(peersIds));
   }
 
   static Available(coValue: CoValueCore) {
-    return new CoValueState(new CoValueAvailableState(coValue));
+    return new CoValueState(coValue.id, new CoValueAvailableState(coValue));
   }
 
-  dispatch(action: CoValueStateAction) {
+  async loadFromPeers(peers: PeerState[]) {
```

Comment:
Moved `loadFromPeers` from the SyncManager to here.

This way all the loading state management is self-contained inside `CoValueState` and it is easier to cover with tests.